stackslib/src/net/neighbors/db.rs

```diff
@@ -0,0 +1,547 @@
+        db_data: Option<&StackerDBHandshakeData>,
+    ) -> Result<Neighbor, net_error>;
+
+    /// Update the given neighbor with optional new handshake state and save it to the DB.
+    /// Returns the updated neighbor.
+    fn update_neighbor(
+        &self,
+        network: &mut PeerNetwork,
+        cur_neighbor: Neighbor,
+        new_data: Option<&HandshakeAcceptData>,
+        new_db_data: Option<&StackerDBHandshakeData>,
+    ) -> Result<Neighbor, net_error>;
+
+    /// Get the number of peers in a given AS
+    fn get_asn_count(&self, network: &PeerNetwork, asn: u32) -> u64;
+
+    /// Get a random starting neighbor for an ongoing walk.
+    /// Older but still fresh neighbors will be preferred -- a neighbor from the first 50th
+    /// percentile of neighbors (by last contact time) will be selected at random.
+    /// Returns the random neighbor on success
+    /// Returns NoSuchNeighbor if there are no candidates
+    fn get_next_walk_neighbor(&self, network: &PeerNetwork) -> Result<Neighbor, net_error> {
+        // pick a random neighbor as a walking point.
+        // favor neighbors with older last-contact times
+        let mut next_neighbors = self
+            .get_fresh_random_neighbors(network, (NUM_NEIGHBORS as u64) * 2)
+            .map_err(|e| {
+                debug!(
+                    "{:?}: Failed to load initial walk neighbors: {:?}",
+                    network.get_local_peer(),
+                    &e
+                );
+                e
+            })?;
+
+        if next_neighbors.len() == 0 {
+            return Err(net_error::NoSuchNeighbor);
+        }
+
+        next_neighbors.sort_by(|n1, n2| n1.last_contact_time.cmp(&n2.last_contact_time));
+        let median_neighbor_idx = next_neighbors.len() / 2;
+        let random_neighbor_idx = if median_neighbor_idx > 0 {
+            thread_rng().gen::<usize>() % median_neighbor_idx
+        } else {
+            0
+        };
+
+        Ok(next_neighbors[random_neighbor_idx].clone())
+    }
+}
+
+/// Production database I/O implementation that uses PeerDB
+pub struct PeerDBNeighborWalk {}
+
+/// Database I/O helpers for the NeighborWalkDB implementation
+impl PeerDBNeighborWalk {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    /// Given a neighbor we tried to insert into the peer database, find one of the existing
+    /// neighbors it collided with.  Return its slot in the peer db.
+    fn find_replaced_neighbor_slot(
+        conn: &DBConn,
+        nk: &NeighborKey,
+    ) -> Result<Option<u32>, net_error> {
+        let mut slots = PeerDB::peer_slots(conn, nk.network_id, &nk.addrbytes, nk.port)
+            .map_err(net_error::DBError)?;
+
+        if slots.len() == 0 {
+            // not present
+            return Ok(None);
+        }
+
+        let mut rng = thread_rng();
+        slots.shuffle(&mut rng);
+        Ok(Some(slots[0]))
+    }
+}
+
+impl NeighborWalkDB for PeerDBNeighborWalk {
+    /// implements get_fresh_random_neighbors
```

Comment:
Are these "implements ..." comments really helpful? Are they just meant to be placeholders?

Reply:
Will remove -- they're an artifact from an earlier version of this code